### PR TITLE
example: add dspy demo build with tidb vector.

### PR DESCRIPTION
**Pull request by AntiKnot**
### What problem does this PR solve?

Issue Number: open #44 

Problem Summary:

### What changed and how does it work?
Add an example demonstrates how to use the DSPy and TiDB Serverless to build a simple RAG application.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by AntiKnot** [comment]
compile failed.
```
compiled_rag = teleprompter.compile(rag, trainset=trainset)
```
will raise 
```
TypeError: cannot pickle 'module' object
```

debug
```
import pickle

def find_non_pickleable_attributes(obj):
    non_pickleable_attrs = {}
    for attr_name, attr_value in vars(obj).items():
        try:
            pickle.dumps(attr_value)
        except pickle.PickleError as e:
            non_pickleable_attrs[attr_name] = str(e)
    return non_pickleable_attrs

#  student is object
non_pickleable_attrs = find_non_pickleable_attributes(student)
if non_pickleable_attrs:
    print("can't pickle attr are：")
    for attr, error in non_pickleable_attrs.items():
        print(f"attr：{attr}, error：{error}")
else:
    print("all attr can pickle")
```
`self._bind = self._create_engine()` sqlalchemy.connect can't pickle.
manually change TiDBVectorClient `def __deepcopy__()` magic feature. or singleton can fix.

File: examples/dspy-demo/utils.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        query_embeddings = self.embedding_function(query_or_queries)
+        k = k or self.top_k
+        tidb_vector_res = self.tidb_vector_client.query(query_vector=query_embeddings, k=k)
+        passages_scores = {}
+        for res in tidb_vector_res:
+            res.metadata = dotdict(res.metadata)
+            passages_scores[res.document] = res.distance
+        sorted_passages = sorted(passages_scores.items(), key=lambda x: x[1], reverse=True)
+        k_sorted_passages = sorted_passages[:k]
```
Comment (IANTHEREAL) [inline examples/dspy-demo/utils.py:98]: Do we need to select the top K items again after invoking `self.tidb_vector_client.query(query_vector=query_embeddings, k=k)`?

File: examples/dspy-demo/utils.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+from typing import Union, List, Optional
+import dspy
+from sentence_transformers import SentenceTransformer
+from tidb_vector.integrations import TiDBVectorClient
+
+
+def sentence_transformer_embedding_function(
+        embed_model: SentenceTransformer,
+        sentences: Union[str, List[str]]
+) -> List[float]:
```
Comment (IANTHEREAL) [inline examples/dspy-demo/utils.py:11]: return `List[List[float]]`?

**Review by IANTHEREAL — COMMENTED** [review]
Thank you for your contribution! It must be very useful.
We hope to move on and  merge this example quickly.

**Comment by IANTHEREAL** [comment]
> ### TiDBVectorClient as RM cannot be deepcopy during compile
> compile failed.
> 
> ```
> compiled_rag = teleprompter.compile(rag, trainset=trainset)
> ```
> 
> will raise
> 
> ```
> TypeError: cannot pickle 'module' object
> ```


Could you give us a favor to fix it?  @AntiKnot

File: examples/dspy-demo/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
+# DSPy Demo
+
+This example demonstrates how to use the DSPy and TiDB Serverless to build a simple RAG application.
+
+## Prerequisites
+
+- A running TiDB Serverless cluster
+    - Vector search enabled
+- Python 3.10 or later
+- Ollama or OpenAI
```
[the remark this answers is not in the document]
Comment (AntiKnot) [reply examples/dspy-demo/README.md:9]: Yes, It support OpenAI API.  
DSPy provides extensive support for the LM API.

Ollama 
```
language_model = dspy.OllamaLocal(...)
```
OpenAI
```
language_model = dspy.OpenAI(...)
```

Refs.
https://dspy-docs.vercel.app/api/category/language-model-api-clients
https://dspy-docs.vercel.app/api/category/local-language-model-clients

File: examples/dspy-demo/utils.py (hunk 1 above)
Comment (AntiKnot) [reply examples/dspy-demo/utils.py:98]: Not necessary, the parameter k in the code snippet has no effect. 
I will remove it.

File: examples/dspy-demo/utils.py (hunk 2 above)
Comment (AntiKnot) [reply examples/dspy-demo/utils.py:11]: I need to first verify that model embeding handles two different signature types: List[str] and List[List[str]]. 
In addition, tidb_vector_client.query supports signature vector but not vectors.

I need to consider how to deal with it.

**Comment by AntiKnot** [comment]
> Could you give us a favor to fix it? @AntiKnot

okay, let me try.

File: tidb_vector/integrations/vector_client.py [hunk 4]
```diff
@@ -175,6 +176,22 @@ def __del__(self) -> None:
...
+
+        # Copy all attributes except the engine connection (_bind)
+        for k, v in self.__dict__.items():
+            if k != "_bind":  # Skip copying the engine connection
+                setattr(result, k, copy.deepcopy(v, memo))
+
+        # Directly assign the engine connection without copying
+        result._bind = self._bind
+
+        return result
```
Comment (AntiKnot) [inline tidb_vector/integrations/vector_client.py:193]: @IANTHEREAL fix TiDBVectorClient as RM cannot be deepcopy during compile

**Review by IANTHEREAL — APPROVED** [review]
LGTM

**Comment by IANTHEREAL** [comment]
@wd0517 should we skip the testing?

**Comment by wd0517** [comment]
> @wd0517 should we skip the testing?

Our tests does not cover example code, it's ok to skip.

**Comment by IANTHEREAL** [comment]
@AntiKnot Thanks for contribution

**Pull request merged**